### DISCUSSION TITLE
Use Apicurio Schemaregistry in Confluent kafka

### DISCUSSION THREAD

**Discussion by ibalachandar86** [Apicurio/apicurio-registry#2459] [Q&A]
Hi,

I am using confluentinc/cp-kafka-connect docker image to run kafka connectors, using confluent schema registry.
Only Kafka connectors are running in Kubernetes not Kafka brokers and Schema registry. Below is my working Kubernetes yaml file with confluent schema registry. 

# Kubernetes YAML file
________________________________________________________
cp-control-center:
  enabled: false
cp-kafka:
  enabled: false
cp-kafka-rest:
  enabled: false
cp-ksql-server:
  enabled: false
cp-schema-registry:
  enabled: false
cp-zookeeper:
  enabled: false
cp-kafka-connect:
  replicaCount: 1

  image: {confluentinc/cp-kafka-connect image with kafka connectors}
  imageTag: XXXX
  imagePullPolicy: IfNotPresent
  servicePort: 8083

>   cp-schema-registry:
>     url: "https://XXXX:8081"

  configurationOverrides:

>     "schema.registry.ssl.truststore.location": "/home/appuser/kafka.truststore.jks"
>     "schema.registry.ssl.truststore.password": "XXXX"
>     "key.converter.schemas.enable": "false"
>     "value.converter.schemas.enable": "false"
>     "basic.auth.credentials.source": "XXXX"

    "config.storage.topic": "_kafka-connect-hdfs-source-connector-configs"
    "offset.storage.topic": "_kafka-connect-hdfs-source-connector-offsets"
    "status.storage.topic": "_kafka-connect-hdfs-source-connector-status"
    "plugin.path": "/usr/share/java,/usr/share/confluent-hub-components,/etc/kafka-connect/jars"
    "key.converter": "io.confluent.connect.avro.AvroConverter"
    "value.converter": "io.confluent.connect.avro.AvroConverter"
    "internal.key.converter": "org.apache.kafka.connect.json.JsonConverter"
    "internal.value.converter": "org.apache.kafka.connect.json.JsonConverter"
    "security.protocol": "SASL_SSL"
    "sasl.mechanism": "GSSAPI"
    "sasl.jaas.config": com.sun.security.auth.module.Krb5LoginModule required \
      useTicketCache=false \
      renewTGT=false \
      useKeyTab=true \
      keyTab="/home/appuser/ssl/svc-dev.keytab" \
      principal="svc-dev";
    "sasl.kerberos.service.name": "kafka"
    "ssl.truststore.location": "/home/appuser/kafka.truststore.jks"
    "ssl.truststore.password": "XXXX"

  kafka:
    bootstrapServers: "https://XXXX:9092" 
________________________________________________________

Instead of Confluent Schema registry, We are trying to use Apicurio schema registry. We have installed Apicurio SR in a separate server not in Kubernetes.
In above YAML file, We have to provide Apicurio schema registry configs in place of quoted Confluent schema registry properties.
I could able to create tokens for my credentials and execute get/post schemas using curl command in APICurio SR.

My questions are,
1) Is it possible to use Apicurio schema registry in combination with Confluent Kafka?

2) If yes, what are the changes/configs I have use to make it work.

3) How to convert this token into authentication properties/configs?

**Comment by Ganesh-403** [comment]
Yes, it is possible to use **Apicurio Registry with a Confluent Kafka deployment**. The Schema Registry is independent of the Kafka brokers, so your Kafka Connect workers can continue talking to Confluent Kafka while using Apicurio Registry for schema management.

The main question is **which client libraries you're using**.

### Option 1: Keep the Confluent converters (recommended for your setup)

Since you're using:

```properties
key.converter=io.confluent.connect.avro.AvroConverter
value.converter=io.confluent.connect.avro.AvroConverter
```

you don't need to replace the converters. Instead, point them at Apicurio Registry's **Confluent compatibility API** rather than the native Apicurio API.

For example:

```properties
schema.registry.url=https://<apicurio-host>/apis/ccompat/v7
```

(or `/apis/ccompat/v8` on newer versions).

Apicurio implements the Confluent Schema Registry REST API specifically for this use case, so Confluent clients can work with Apicurio Registry by simply changing the registry URL. Avro, JSON Schema, and Protobuf are all supported through this compatibility layer. :contentReference[oaicite:0]{index=0}

Your converter configuration would therefore remain largely unchanged:

```properties
key.converter=io.confluent.connect.avro.AvroConverter
value.converter=io.confluent.connect.avro.AvroConverter

key.converter.schema.registry.url=https://<apicurio-host>/apis/ccompat/v7
value.converter.schema.registry.url=https://<apicurio-host>/apis/ccompat/v7
```

### Authentication

If you've already verified that you can access the Registry using a bearer token via `curl`, the equivalent Kafka Connect configuration depends on how your Apicurio Registry is secured.

For an OIDC-protected Registry, the Confluent clients typically use the standard Schema Registry authentication properties (for example, bearer authentication or basic authentication if configured). The exact properties depend on your authentication mechanism rather than Apicurio itself. Apicurio's compatibility layer is designed to work with Confluent clients while using Apicurio's authentication mechanisms. :contentReference[oaicite:1]{index=1}

### Option 2: Switch to Apicurio converters

Alternatively, you could replace the Confluent converters with the Apicurio Kafka Connect converters, but that isn't necessary if your goal is simply to replace the Schema Registry while keeping the rest of your Confluent-based deployment.

### A couple of questions

To give more specific configuration, could you clarify:

1. Which version of **Apicurio Registry** are you running?
2. Is your Registry secured with **OIDC/Keycloak**, **Basic Authentication**, or another authentication mechanism?
3. Are you planning to continue using the **Confluent AvroConverter**, or are you considering switching to the Apicurio converters?

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily. @ibalachandar86
